network/hostport.go

```diff
@@ -276,77 +247,39 @@ type Dialer interface {
-// DialResult holds the results written by DialHostPort() on its returned
-// channel.
-type DialResult struct {
-	Endpoint *HostPort
-	Duration time.Duration
-	Error    error
-}
-
-// DialHostPort dials the given hostPort asynchronously and writes a DialResult
-// on the returned channel when done, then closing the channel. If a connection
-// was successfully established, it will be closed immediately. The given dialer
-// and clock are used to perform the dialing and measuring the time taken (which
-// is also set on error). Usually, a net.Dialer and clock.WallClock are passed
-// for dialer and clock, respectively.
-func DialHostPort(hostPort HostPort, dialer Dialer, clock clock.Clock) <-chan DialResult {
-	dialResult := make(chan DialResult, 1)
-
-	go func() {
-		defer close(dialResult)
-
-		result := DialResult{
-			Endpoint: &hostPort,
-		}
-
-		startTime := clock.Now()
-
-		conn, err := dialer.Dial("tcp", hostPort.NetAddr())
-
-		result.Duration = clock.Now().Sub(startTime)
-		if err != nil {
-			result.Error = err
-		}
-
-		if conn != nil {
-			conn.Close()
-		}
-
-		dialResult <- result
-	}()
-
-	return dialResult
+type dialResult struct {
+	endpoint HostPort
+	err      error
 }
 
-// FastestHostPort uses the given dialer and clock to determine the best
-// HostPort from the given hostPorts. Each unique HostPort entry is dialed in
-// parallel, measuring the time taken to establish a successful connection,
-// which is then closed. Individual connection errors are discarded, and an
-// error is returned only if none of the hostPorts can be reached.
-func FastestHostPort(dialer Dialer, clock clock.Clock, hostPorts []HostPort) (HostPort, error) {
-	stop := make(chan struct{})
-	defer close(stop)
-
-	bestResult := DialResult{Duration: 24 * time.Hour}
-	for hostPort := range UniqueHostPorts(stop, hostPorts) {
-		lastResult := <-DialHostPort(hostPort, dialer, clock)
-
-		if lastResult.Error != nil {
-			// Discard individual connection errors.
-			logger.Debugf("discarding error dialing %q: %v", lastResult.Endpoint, lastResult.Error)
-			continue
-		}
-
-		if bestResult.Duration > lastResult.Duration {
-			bestResult = lastResult
-			logger.Debugf("best endpoint %q has duration %s", bestResult.Endpoint, bestResult.Duration)
-		}
+// FastestHostPort dials the unique entries in the given hostPorts, in parallel,
+// using the given dialer, closing successfully established connections
+// immediately. Individual connection errors are discarded, and an error is
+// returned only if none of the hostPorts can be reached when the given timeout
+// expires.
+//
+// Usually, a net.Dialer initialized with a non-empty Timeout field is passed
+// for dialer.
+func FastestHostPort(hostPorts []HostPort, dialer Dialer, timeout time.Duration) (HostPort, error) {
+	uniqueHPs := UniqueHostPorts(hostPorts)
+	successful := make(chan HostPort, 1)
+
+	for _, hostPort := range uniqueHPs {
+		go func(hp HostPort) {
+			conn, err := dialer.Dial("tcp", hp.NetAddr())
```

Comment:
It's not immediately clear why we can Dial without a timeout. Is it because of the select/timeout further down? And if some probes take longer than others do these active go routines close all their connections?